include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -1827,282 +1151,559 @@ namespace internal
+      typename FEEvaluationData<dim, Number, false>::shape_info_number_type;
+    using Eval = EvaluatorTensorProduct<evaluate_evenodd,
+                                        dim,
+                                        fe_degree + 1,
+                                        fe_degree + 1,
+                                        Number,
+                                        Number2>;
+
+    static void
+    evaluate(const unsigned int                     n_components,
+             const EvaluationFlags::EvaluationFlags evaluation_flag,
+             const Number                          *values_dofs,
+             FEEvaluationData<dim, Number, false>  &fe_eval)
+    {
+      constexpr std::size_t n_points = Utilities::pow(fe_degree + 1, dim);
+
+      for (unsigned int c = 0; c < n_components; ++c)
+        {
+          if ((evaluation_flag & EvaluationFlags::values) != 0u)
+            for (unsigned int i = 0; i < n_points; ++i)
+              fe_eval.begin_values()[n_points * c + i] =
+                values_dofs[n_points * c + i];
+
+          if ((evaluation_flag & EvaluationFlags::gradients) != 0u)
+            evaluate_gradients_collocation<fe_degree + 1, dim>(
+              fe_eval.get_shape_info().data.front(),
+              values_dofs + c * n_points,
+              fe_eval.begin_gradients() + c * dim * n_points);
+        }
+    }
+
+    static void
+    integrate(const unsigned int                     n_components,
+              const EvaluationFlags::EvaluationFlags integration_flag,
+              Number                                *values_dofs,
+              FEEvaluationData<dim, Number, false>  &fe_eval,
+              const bool                             add_into_values_array)
+    {
+      constexpr std::size_t n_points = Utilities::pow(fe_degree + 1, dim);
+
+      for (unsigned int c = 0; c < n_components; ++c)
+        {
+          if ((integration_flag & EvaluationFlags::values) != 0u)
+            {
+              if (add_into_values_array)
+                for (unsigned int i = 0; i < n_points; ++i)
+                  values_dofs[n_points * c + i] +=
+                    fe_eval.begin_values()[n_points * c + i];
+              else
+                for (unsigned int i = 0; i < n_points; ++i)
+                  values_dofs[n_points * c + i] =
+                    fe_eval.begin_values()[n_points * c + i];
+            }
+
+          if ((integration_flag & EvaluationFlags::gradients) != 0u)
+            integrate_gradients_collocation<fe_degree + 1, dim>(
+              fe_eval.get_shape_info().data.front(),
+              values_dofs + c * n_points,
+              fe_eval.begin_gradients() + c * dim * n_points,
+              add_into_values_array ||
+                ((integration_flag & EvaluationFlags::values) != 0u));
+        }
+    }
+  };
+
+
+
   /**
    * This struct performs the evaluation of function values, gradients and
    * Hessians for tensor-product finite elements. This a specialization for
```

Comment:
```suggestion
   * This struct performs the evaluation of function values and gradients
   * for tensor-product finite elements. This a specialization for
```
Hessians are done in the Selector now, not in the specialised structs. We should mention this!